### ISSUE TITLE
Formatting negative money amounts with ISO symbol may have incorrect spacing

### ISSUE THREAD

**Issue by kipcole9** [kipcole9/money#147]
Per @rubas's comment in Elixir Slack:

![screenshot_2023-02-22_at_14 22 59_720](https://user-images.githubusercontent.com/15895/221028371-a7d3bdf3-620e-4316-872f-5ca8bd4b6c5a.jpg)

**Comment by kipcole9** [comment]
The relevant part of [TR35](https://unicode.org/reports/tr35/tr35-numbers.html#4-currencies) doesn't cater for negative amounts when using "currency spacing" which is what `ex_cldr_number` uses today:

> When a currency symbol is substitited into a pattern, some spacing adjustments or other adjustments may be necessary depending on the nature of the symbol. In CLDR 42 and later, the preferred way to handle this is via the alt="alphaNextToNumber" variant of the currencyFormat pattern, as described in [Section 2.4.2: Currency Formats](https://unicode.org/reports/tr35/tr35-numbers.html#Currency_Formats). In earlier versions of CLDR this was handled via the currencySpacing element as described below. This element is still present in CLDR 42 and its use is described below for implementations that may not yet support the alt="alphaNextToNumber" variant of the currencyFormat pattern.

```xml
<currencySpacing>
  <beforeCurrency>
    <currencyMatch>[:^S:]</currencyMatch>
    <surroundingMatch>[:digit:]</surroundingMatch>
    <insertBetween> </insertBetween>
  </beforeCurrency>
  <afterCurrency>
    <currencyMatch>[:^S:]</currencyMatch>
    <surroundingMatch>[:digit:]</surroundingMatch>
    <insertBetween> </insertBetween>
  </afterCurrency>
</currencySpacing>
```

> This element controls whether additional characters are inserted on the boundary between the symbol and the pattern. For example, with the above currencySpacing, inserting the symbol "US$" into the pattern "#,##0.00¤" would result in an extra no-break space inserted before the symbol, for example, "#,##0.00 US$". The beforeCurrency element governs this case, since we are looking before the "¤" symbol. The currencyMatch is positive, since the "U" in "US$" is at the start of the currency symbol being substituted. The surroundingMatch is positive, since the character just before the "¤" will be a digit. Because these two conditions are true, the insertion is made.

> Conversely, look at the pattern "¤#,##0.00" with the symbol "US$". In this case, there is no insertion; the result is simply "US$#,##0.00". The afterCurrency element governs this case, since we are looking after the "¤" symbol. The surroundingMatch is positive, since the character just after the "¤" will be a digit. However, the currencyMatch is not positive, since the "$" in "US$" is at the end of the currency symbol being substituted. So the insertion is not made.

**Comment by kipcole9** [comment]
The current advise is to use the `alphaNextToNumber` format:

> The alt="alphaNextToNumber" pattern, if available, should be used instead of the standard pattern when the currency symbol character closest to the numeric value has Unicode General Category L (letter). The alt="alphaNextToNumber" pattern is typically provided when the standard currency pattern does not have a space between currency symbol and numeric value; the alphaNextToNumber variant adds a non-breaking space if appropriate for the locale.

Those formats look like:
* `en` => ` currency_alpha_next_to_number: "¤ #,##0.00"`
* `de-CH` => currency_alpha_next_to_number: "¤ #,##0.00"

Hence I believe changing the formatter to use these formats will resolve the issue (without any change to your code).  Thankfully I have a long flight coming up this weekend so I have time to work on it then and aim to have a new release out by Monday.

**Comment by kipcole9** [comment]
This is taking a little longer than expected and will require a bit of refactoring of the format parser (in order to prevent a performance degradation when I add the new format resolver).  Still aiming for end of the weekend but I'm also on a 12 hour flight on Sunday so it depends on how much sleep I need :-)

**Comment by kipcole9** [comment]
@rubas, finally finished this up and ready for your to review if you wouldn't mind.  If you put `{:ex_cldr_numbers, github: "elixir-cldr/cldr_numbers", override: true}` in `mix.exs` and then `mix deps.get` I think you should be good to go.

This seemingly small change needed quite a lot of refactoring so although all tests pass, there is always the possibility of some unforeseen incompatibility.

Before I publish this version I will also add a mechanism to set a default currency format (either `:currency` or `:accounting`).

**Comment by kipcole9** [comment]
I've merged the required changed into the master branch so for testing, just `{:ex_cldr_numbers, github: "elixir-cldr/cldr_numbers", override: true}` is required now.

I have added an option `:default_currency_format` to the Cldr backend configuration options. It can be set to `:currency`, `:accounting` or `nil`. Not applicable to this issue, just an FYI.

Let me know your feedback if any and I'll publish the update.

**Comment by kipcole9** [comment]
[ex_cldr_numbers version 2.30.0](https://hex.pm/packages/ex_cldr_numbers/2.30.0) has been published with the following changelog entries. I believe this resolves your issue but of course please re-open if not.

### Bug Fixes

* Fix short formatting for decimal numbers. Thanks to @dbernheisel for the report. Closes #34.

* Fix formatting negative numbers when the currency symbol results in an alphabetic character next to the number. In these cases a special format `:currency_alpha_next_to_number` (or `:accounting_alpha_next_to_number`) is used. These formats are only available in CLDR42 or later and therefore `ex_cldr` version 2.35.0 or later is required.

### Enhancements

* Supports the `:default_currency_format` backend configuration option introduced in `ex_cldr` version 2.35.0.

* Adds `:from_locale` as a valid value to the `:currency` option to `Cldr.Number.to_string/3`. If set, the currency is derived from the locale.
